torchtitan/experiments/llama4/model/state_dict_adapter.py

```diff
@@ -0,0 +1,167 @@
+                key = re.sub(r"(\d+)", "{}", key, count=1)
+            else:
+                layer_num = None
+
+            if key in to_hf_map:
+                # do direct mapping
+                if key in "layers.{}.moe.experts.w2":
+                    value = value.transpose(-1, -2)
+
+                new_key = to_hf_map[key]
+                if new_key is None:
+                    continue
+                if layer_num:
+                    new_key = new_key.format(layer_num)
+                hf_state_dict[new_key] = value
+            elif key in self.reverse_combination_plan:
+                # handle collecting values to combine
+                hf_abstract_key = self.reverse_combination_plan[key]
+                if hf_abstract_key is None:
+                    continue
+                to_combine[hf_abstract_key.format(layer_num)][
+                    key.format(layer_num)
+                ] = value
+
+        # combine collected values
+        for hf_fqn, tt_fqn_map in to_combine.items():
+            layer_num = re.search(r"\d+", hf_fqn).group(0)
+            hf_abstract_key = re.sub(r"(\d+)", "{}", hf_fqn, count=1)
+            combine_values = []
+            # use combination_plan to ensure correct order before concatenation
+            for tt_abstract_key in self.combination_plan[hf_abstract_key]:
+                tt_key = tt_abstract_key.format(layer_num)
+                print("tt_key", tt_key, "shape", tt_fqn_map[tt_key].shape)
+                combine_values.append(tt_fqn_map[tt_key].transpose(-1, -2))
+
+            value = torch.cat(combine_values, dim=-1)
+            hf_state_dict[hf_fqn] = value
+
+        return hf_state_dict
+
+    def from_hf(self, hf_state_dict: dict[str, Any]) -> dict[str, Any]:
+        state_dict = {}
+
+        for key, value in hf_state_dict.items():
+            if "layers" in key:
+                layer_num = re.search(r"\d+", key).group(0)
+                key = re.sub(r"(\d+)", "{}", key, count=1)
+            else:
+                layer_num = None
+
+            if key in self.from_hf_map:
+                # do direct mapping
+                if key in [
+                    "language_model.model.layers.{}.feed_forward.experts.down_proj",
+                ]:
+                    value = value.transpose(-1, -2)
+
+                new_key = self.from_hf_map[key]
+                if new_key is None:
+                    continue
+                if layer_num:
+                    new_key = new_key.format(layer_num)
+                state_dict[new_key] = value
+            elif key in [
+                "language_model.model.layers.{}.feed_forward.experts.gate_up_proj"
+            ]:
+                # handle splitting values
+                split_vals = value.chunk(2, dim=-1)
+                split_vals = [val.transpose(-1, -2) for val in split_vals]
+                for new_key, split_val in zip(self.combination_plan[key], split_vals):
```

Comment:
So the first half is w1 and the second half is w3? I was thinking we could even remove the self.combination_plan, and directly hardcode the split/concate and the corresponding torchtitan fqn in from_hf(). This is because we only have 2 titan fqns to split, not too much .Now we still need to hard code the abstract_key (eg, "language_model.model.layers.{}.feed_forward.experts.gate_up_proj").